packages/next-swc/crates/napi/src/next_api/project.rs

```diff
@@ -607,3 +612,122 @@ pub fn project_update_info_subscribe(
     });
     Ok(())
 }
+
+#[turbo_tasks::value]
+#[derive(Debug)]
+#[napi(object)]
+pub struct TracedSource {
+    pub frame: StackFrame,
+    pub source: String,
+}
+
+#[turbo_tasks::value]
+#[derive(Debug)]
+#[napi(object)]
+pub struct StackFrame {
+    pub file: String,
+    pub method_name: Option<String>,
+    pub line: u32,
+    pub column: Option<u32>,
+}
+
+#[napi]
+pub async fn project_trace_source(
+    #[napi(ts_arg_type = "{ __napiType: \"Project\" }")] project: External<ProjectInstance>,
+    frame: StackFrame,
+) -> napi::Result<Option<TracedSource>> {
+    let turbo_tasks = project.turbo_tasks.clone();
+    let traced_frame = turbo_tasks
+        .run_once(async move {
+            let file = match Url::parse(&frame.file) {
+                Ok(url) => match url.scheme() {
+                    "file" => url.path().to_string(),
+                    _ => bail!("Unknown url scheme"),
+                },
+                Err(_) => frame.file.to_string(),
+            };
+
+            let Some(chunk_base) = file.strip_prefix(
+                &(format!(
+                    "{}/{}",
+                    project.container.project().await?.root_path,
+                    project.container.project().dist_dir().await?
+                )),
+            ) else {
+                // File doesn't exist within the dist dir
+                return Ok(None);
+            };
+
+            let chunk_path = format!(
+                "{}{}",
+                project
+                    .container
+                    .project()
+                    .client_relative_path()
+                    .await?
+                    .path,
+                chunk_base
+            );
+
+            let path = project
+                .container
+                .project()
+                .client_root()
+                .fs()
+                .root()
+                .join(chunk_path);
+
+            let generatable: Vc<Box<dyn GenerateSourceMap>> =
+                Vc::try_resolve_sidecast(project.container.get_versioned_content(path))
+                    .await?
+                    .context("Chunk cannot produce a sourcemap")?;
```

Comment:
Can we return a `None` here instead of erroring out? Not every output will have a `GenerateSourceMap` impl.